src/legacy/huffman.rs

```diff
@@ -0,0 +1,215 @@
+            debug_assert!(sym <= u16::MAX as usize);
+            self.table[index].sym = sym as u16;
+            debug_assert!(len <= u8::MAX as usize);
+            self.table[index].len = len as u8;
+        }
+    }
+
+    /// Use the decoder d to decode a symbol from the LSB-first zero-padded bits.
+    /// Returns the decoded symbol number or -1 if no symbol could be decoded.
+    /// *num_used_bits will be set to the number of bits used to decode the symbol,
+    /// or zero if no symbol could be decoded.
+    pub fn huffman_decode(&mut self, bits: u16, num_used_bits: &mut u8) -> std::io::Result<u16> {
+        // First try the lookup table.
+        let lookup_bits = lsb(bits as u64, HUFFMAN_LOOKUP_TABLE_BITS) as usize;
+        debug_assert!(lookup_bits < self.table.len());
+
+        if self.table[lookup_bits].len != 0 {
+            debug_assert!(self.table[lookup_bits].len < HUFFMAN_LOOKUP_TABLE_BITS);
+            //  debug_assert!(self.table[lookup_bits].sym < self.num_syms);
+            *num_used_bits = self.table[lookup_bits].len;
+            return Ok(self.table[lookup_bits].sym);
+        }
+
+        // Then do canonical decoding with the bits in MSB-first order.
+        let mut bits = reverse_lsb(bits, MAX_HUFFMAN_BITS);
+        for l in HUFFMAN_LOOKUP_TABLE_BITS as usize + 1..MAX_HUFFMAN_BITS {
+            if self.sentinel_bits[l] > bits as u32 {
+                bits >>= MAX_HUFFMAN_BITS - l;
+
+                let sym_idx = self.offset_first_sym_idx[l] + bits;
+                // debug_assert(sym_idx < self.num_syms);
+
+                *num_used_bits = l as u8;
+                return Ok(self.syms[sym_idx as usize]);
+            }
+        }
+        *num_used_bits = 0;
+        Err(Error::new(
+            io::ErrorKind::InvalidData,
+            "huffman decode failed",
+        ))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::HuffmanDecoder;
+
+    #[test]
+    fn test_huffman_decode_basic() {
+        let lens = [
+            3, // sym 0:  000
+            3, // sym 1:  001
+            3, // sym 2:  010
+            3, // sym 3:  011
+            3, // sym 4:  100
+            3, // sym 5:  101
+            4, // sym 6:  1100
+            4, // sym 7:  1101
+            0, // sym 8:
+            0, // sym 9:
+            0, // sym 10:
+            0, // sym 11:
+            0, // sym 12:
+            0, // sym 13:
+            0, // sym 14:
+            0, // sym 15:
+            6, // sym 16: 111110
+            5, // sym 17: 11110
+            4, // sym 18: 1110
+        ];
+
+        let mut d = HuffmanDecoder::default();
+        d.init(&lens, lens.len()).unwrap();
+
+        let mut used = 0;
+        // 000 (msb-first) -> 000 (lsb-first)
+        assert_eq!(d.huffman_decode(0x0, &mut used).unwrap(), 0);
+        assert_eq!(used, 3);
+
+        /* 011 (msb-first) -> 110 (lsb-first)*/
+        assert_eq!(d.huffman_decode(0x6, &mut used).unwrap(), 3);
```

Comment:
Again, binary literals.

Reply:
here it makes sense - esp. with the comments.